src/Lecture4.hs

```diff
@@ -131,8 +142,30 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+packRow :: (String, TradeType, Int) -> Row
+packRow (prod, tradeType, cost) = Row { rowProduct = prod, rowTradeType = tradeType, rowCost = cost }
+
+readRow :: [String] -> Maybe Row
+readRow [prodS, tradeTypeS, costS] =
+    if prodS == "" then empty
+    else case tradeTypeM of
+        Nothing        -> empty
+        Just tradeType ->
+            case costM of
+                Nothing -> empty
+                Just cost ->
+                    if cost < 0 then empty
+                    else Just $ packRow (prodS, tradeType, cost)
```

Comment:
`empty` is `Nothing` for `Maybe`. Returning `Nothing` on `Nothing` and doing something with the value inside `Just` is exactly the `Monad` instance for `Maybe`! So you can simplify this function by using the `>>=` operator or even `do`-notation 🙂